app/models/assignment.rb

```diff
@@ -600,6 +600,59 @@ def summary_json(user)
       numMarked: self.get_num_marked(user.admin? ? nil : user.id) }
   end
 
+  # Generates the summary of the most test results associated with an assignment.
+  def summary_test_results
+    test_groups_query = self.test_groups
+                            .joins(:test_group_results)
+                            .group(:id)
+                            .select('test_groups.id AS test_groups_id',
+                                    'MAX(test_group_results.created_at) AS test_group_results_created_at')
+                            .to_sql
+    test_group_query = TestGroupResult.joins(:test_group)
+                                      .joins("INNER JOIN (#{test_groups_query}) sub \
+                                              ON test_groups.id = sub.test_groups_id \
+                                              AND test_group_results.created_at = sub.test_group_results_created_at")
+                                      .select('test_groups.name AS test_group_name', 'test_group_results.*')
+                                      .to_sql
+
+    TestResult.joins("INNER JOIN (#{test_group_query}) sub ON test_results.id = sub.test_run_id")
+              .select('test_results.name as test_result_name',
+                      'test_results.output',
+                      'test_results.marks_earned as test_result_marks_earned',
+                      'test_results.marks_total as test_result_marks_total',
+                      'test_results.id as id',
+                      :status, :marks_earned, :marks_total, :extra_info, :error_type)
+  end
+
+  # Generate a JSON summary of the most recent test results associated with an assignment.
+  def summary_test_result_json
+    self.summary_test_results.to_json
+  end
+
+  # Generate a CSV summary of the most recent test results associated with an assignment.
+  def summary_test_result_csv
+    summary_test_result = self.summary_test_results
+    CSV.generate do |csv|
+      submissions = {}
+      summary_test_result.map do |test_result|
```

Comment:
This approach is pretty good, but has two issues:

1. It assumes that the test results are well-ordered, so that results for each grouping will be added to `submissions` in the same order.
2. It doesn't handle the case when a grouping is missing some results---logically this should create a "gap" in a row, but instead the row's other elements will be shifted over.

So I'll suggest a more explicit creation of a table, by first querying all distinct test result names from `summary_test_result`, then grouping the results by grouping id (similar to what I suggested in the JSON method), and for each grouping constructing the correct row by cross-referencing the results against the collection of test result names. I'll leave the details of that part for you to work out.